File: app/src/main/java/org/lineageos/updater/ExportUpdateService.java
/*
 * Copyright (C) 2017-2022 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.updater;

import android.app.Service;
import android.content.ContentResolver;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.net.Uri;
import android.os.IBinder;
import android.os.SystemClock;
import android.util.Log;
import android.widget.Toast;

import org.lineageos.updater.misc.FileUtils;
import org.lineageos.updater.misc.NotificationHelper;

import java.io.File;
import java.io.IOException;

public class ExportUpdateService extends Service {

    private static final String TAG = "ExportUpdateService";

    public static final String ACTION_START_EXPORTING = "start_exporting";

    public static final String EXTRA_SOURCE_FILE = "source_file";
    public static final String EXTRA_DEST_URI = "dest_uri";

    private volatile boolean mIsExporting = false;

    private Thread mExportThread;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (ACTION_START_EXPORTING.equals(intent.getAction())) {
            if (mIsExporting) {
                Log.e(TAG, "Already exporting an update");
                Toast.makeText(this, R.string.toast_already_exporting, Toast.LENGTH_SHORT).show();
                return START_NOT_STICKY;
            }
            mIsExporting = true;
            File source = (File) intent.getSerializableExtra(EXTRA_SOURCE_FILE);
            Uri destination = intent.getParcelableExtra(EXTRA_DEST_URI);
            startExporting(source, destination);
            Toast.makeText(this, R.string.toast_export_started, Toast.LENGTH_SHORT).show();
        } else {
            Log.e(TAG, "No action specified");
        }

        if (!mIsExporting) {
            stopSelf();
        }

        return START_NOT_STICKY;
    }

    private class ExportRunnable implements Runnable {
        private final ContentResolver mContentResolver;
        private final File mSource;
        private final Uri mDestination;
        private final FileUtils.ProgressCallBack mProgressCallBack;
        private final Runnable mRunnableComplete;
        private final Runnable mRunnableFailed;

        private ExportRunnable(ContentResolver cr, File source, Uri destination,
                               FileUtils.ProgressCallBack progressCallBack,
                               Runnable runnableComplete, Runnable runnableFailed) {
            mContentResolver = cr;
            mSource = source;
            mDestination = destination;
            mProgressCallBack = progressCallBack;
            mRunnableComplete = runnableComplete;
            mRunnableFailed = runnableFailed;
        }

        @Override
        public void run() {
            try {
                FileUtils.copyFile(mContentResolver, mSource, mDestination, mProgressCallBack);
                mIsExporting = false;
                if (!mExportThread.isInterrupted()) {
                    Log.d(TAG, "Completed");
                    mRunnableComplete.run();
                } else {
                    Log.d(TAG, "Aborted");
                }
            } catch (IOException e) {
                mIsExporting = false;
                Log.e(TAG, "Could not copy file", e);
                mRunnableFailed.run();
            } finally {
                stopSelf();
            }
        }
    }

    private void startExporting(File source, Uri destination) {
        final String fileName = FileUtils.queryName(getContentResolver(), destination);
        NotificationHelper notificationHelper = NotificationHelper.getInstance(this);

        if (fileName == null) {
            Log.e(TAG, "Failed to resolve file name from URI: " + destination);
            notificationHelper.notifyExportFinished(null);
            stopSelf();
            return;
        }

        FileUtils.ProgressCallBack progressCallBack = new FileUtils.ProgressCallBack() {
            private long mLastUpdate = -1;

            @Override
            public void update(int progress) {
                long now = SystemClock.elapsedRealtime();
                if (mLastUpdate < 0 || now - mLastUpdate > 500) {
                    notificationHelper.notifyExportProgress(progress, fileName);
                    mLastUpdate = now;
                }
            }
        };

        startForeground(NotificationHelper.NOTIFICATION_ID_EXPORT,
                notificationHelper.buildExportStart(fileName),
                ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE);

        Runnable runnableComplete = () -> {
            notificationHelper.notifyExportFinished(fileName);
            stopForeground(STOP_FOREGROUND_DETACH);
        };

        Runnable runnableFailed = () -> {
            notificationHelper.notifyExportFinished(null);
            stopForeground(STOP_FOREGROUND_DETACH);
        };

        ExportRunnable exportRunnable = new ExportRunnable(getContentResolver(), source,
                destination, progressCallBack, runnableComplete, runnableFailed);
        mExportThread = new Thread(exportRunnable);
        mExportThread.start();
    }
}

File: app/src/main/java/org/lineageos/updater/ExportUpdateService.kt
/*
 * SPDX-FileCopyrightText: 2017-2025 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.updater

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.content.pm.ServiceInfo
import android.net.Uri
import android.os.FileUtils
import android.os.IBinder
import android.util.Log
import android.widget.Toast
import androidx.core.app.NotificationCompat
import org.lineageos.updater.misc.Constants.NOTIFICATION_CHANNEL_EXPORT_UPDATE
import org.lineageos.updater.misc.Constants.NOTIFICATION_ID_EXPORT_UPDATE
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

/**
 * Service responsible for exporting an update package to user-selected storage.
 * Uses android.os.FileUtils for direct file descriptor copy.
 * Notifies user on success or failure with destination information.
 */
class ExportUpdateService : Service() {

    companion object {
        private const val TAG = "ExportUpdateService"

        const val ACTION_START_EXPORTING = "start_exporting"
        const val EXTRA_SOURCE_FILE = "source_file"
        const val EXTRA_DEST_URI = "dest_uri"
    }

    @Volatile
    private var isExporting = false

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action != ACTION_START_EXPORTING) {
            Log.e(TAG, "Invalid action")
            stopSelf()
            return START_NOT_STICKY
        }

        if (isExporting) {
            Log.e(TAG, "Already exporting")
            Toast.makeText(this, R.string.toast_already_exporting, Toast.LENGTH_SHORT).show()
            return START_NOT_STICKY
        }

        isExporting = true

        val source = intent.getSerializableExtra(EXTRA_SOURCE_FILE, File::class.java)
        val destination = intent.getParcelableExtra(EXTRA_DEST_URI, Uri::class.java)

        if (source == null || destination == null) {
            Log.e(TAG, "Missing source or destination")
            stopSelf()
            return START_NOT_STICKY
        }

        Thread { exportFile(source, destination) }.start()
        Toast.makeText(this, R.string.toast_export_started, Toast.LENGTH_SHORT).show()
        return START_NOT_STICKY
    }

    private fun exportFile(source: File, destination: Uri) {
        try {
            val pfd = contentResolver.openFileDescriptor(destination, "w")
                ?: throw IOException("Failed to open destination: $destination")

            FileInputStream(source).use { input ->
                FileOutputStream(pfd.fileDescriptor).use { output ->
                    FileUtils.copy(input, output)
                }
            }

            notifyResult(success = true, destination)
        } catch (e: IOException) {
            Log.e(TAG, "Export failed", e)
            notifyResult(success = false, destination)
        } finally {
            isExporting = false
            stopSelf()
        }
    }

    private fun notifyResult(success: Boolean, destination: Uri) {
        val notificationManager = getSystemService(NotificationManager::class.java)
        val channel = NotificationChannel(
            NOTIFICATION_CHANNEL_EXPORT_UPDATE,
            getString(R.string.export_channel_title),
            NotificationManager.IMPORTANCE_LOW
        )
        notificationManager.createNotificationChannel(channel)

        val title = if (success) {
            getString(R.string.notification_export_success)
        } else {
            getString(R.string.notification_export_fail)
        }

        val text = if (success) {
            getString(R.string.notification_export_success) + ": " + destination.toString()
        } else {
            getString(R.string.notification_export_fail)
        }

        val notification: Notification = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_EXPORT_UPDATE)
            .setSmallIcon(R.drawable.ic_system_update)
            .setContentTitle(title)
            .setContentText(text)
            .setStyle(NotificationCompat.BigTextStyle().bigText(text))
            .build()

        startForeground(
            NOTIFICATION_ID_EXPORT_UPDATE,
            notification,
            ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
        )

        notificationManager.notify(NOTIFICATION_ID_EXPORT_UPDATE, notification)
    }
}

